Extension/src/LanguageServer/configurations.ts

```diff
@@ -317,6 +323,9 @@ export class CppProperties {
         } else {
             configuration.includePath = [defaultFolder];
         }
+        // check settings.addNodeAddonIncludePaths?
```

Comment:
Yeah, can you check settings.addNodeAddonIncludePaths and skip the concat -- that would enable dynamic disabling without a reload. For the enabling case, in onDidChangeSettings you can just check for the addNodeAddonIncludePaths settings being enabled and then call promptForReloadWindowDueToSettingsChange .

Reply:
i used `isUnset()` because all the other checks do but am wondering why `!settings.addNodeAddonIncludePaths` might not be valid.

i'm now looking at *which* `onDidChangeSettings()` would be appropriate to modify. any hints appreciated as i don't see any of them that (directly) call `util.promptForReloadWindowDueToSettingsChange()`.

Reply:
IsUnset is used because in TypeScript values like "" also implicitly convert to false in "if" statements. So using `isUnset` is not what you want, since it just checks if the setting doesn't exist.

Looking more closely at the code, `applyDefaultConfigurationValues` maybe be the wrong place for setting this. The reason the vcpkgIncludes is set here is because we wanted it to only appear when users create a new default c_cpp_properties.json, but in your case it sounds like you want it to always be auto-added?

The correct spot may be `updateServerOnFolderSettingsChange` (I'm not 100% sure).

Yeah, we currently call util.promtForReloadWindowDueToSettingsChange after we get a response from cpptools from the setting change. I was referring to `DefaultClient.onDidChangeSettings(event: vscode.ConfigurationChangeEvent, isFirstClient: boolean): { [key: string]: string }`, which has some existing code that checks if certain settings were changed.

Reply:
> but in your case it sounds like you want it to always be auto-added

correct. those paths should always be auto-added. this is where i am working now. it doesn't appear that `updateServerOnFolderSettingsChange()` is where to add the includes (though it's possible that something else requires changing too).

i'll keep debugging and trying to determine how to get the logic working. i ran into the "unbound breakpoint" problem but was able to recover by checking out a previous commit. it took me a bit though.

Reply:
> * i'm thinking of renaming `addNodeAddonIncludePaths` to `readNodeAddonIncludePaths` because it doesn't add them.

That sounds good.

> (mostly allows changing on the fly) if the user adds `nan` or `node-addon-api` to the project on the fly the code doesn't know that and doesn't know to re-execute `addNodeAddonIncludePaths()`. is there a way to get notified when a specific file (the project's package.json) is modified?

Yes, you can add a file watcher on the package.json and react to changes. We do the same thing for c_cpp_properties.json at https://github.com/microsoft/vscode-cpptools/blob/insiders/Extension/src/LanguageServer/configurations.ts#L207 .

In some cases, the file watchers may file, so you could fallback to checking the time stamp in our idle loop processing like was done in https://github.com/microsoft/vscode-cpptools/pull/6748/files for compile_commands.json .
 
> and i need to clean up the `showErrorMessage('error.adding.node.addon.includes')` but wanted to get your feedback on the preferred way to showing errors like this first.
> 
> feedback is appreciated.

I didn't get a chance to finish reviewing the changes today.